electrum/wallet.py

```diff
@@ -814,16 +1011,120 @@ def add_input_info(self, txin):
+        - bip32_derivation
+        """
+        for inp in psbt.input_sections:
+            utxo = inp.non_witness_utxo
+            if not utxo:
+                continue
+            self.receive_tx_callback(utxo.txid(), utxo, TX_HEIGHT_UNCONFIRMED)
+
+        tx = psbt.glob.unsigned_tx
+        full_utxos = [self.transactions.get(utxo['prevout_hash']) for utxo in tx.inputs()]
+        full_utxos = [tx_to_immutable(tx) if tx else None for tx in full_utxos]  # type: List[ImmutableTransaction]
+
+        # adding bip32_derivation, non_witness_utxo to inputs
+        sig_info = self.get_input_info_psbt(psbt=psbt)
+        for i, utxo in enumerate(full_utxos):
+            if not utxo:
+                continue
+            inp = psbt.input_sections[i]
+            txin = tx.inputs()[i]  # corresponding global txin
+            if is_segwit_input(txin):
+                witness = utxo.get_witness(txin['prevout_n'])
+                if witness != '00':
+                    sig_info[i]['witness_utxo'] = witness
+            else:
+                sig_info[i]['non_witness_utxo'] = utxo
+            sig_info[i]['sighash_type'] = SIGHASH_ALL
+        psbt.update_inputs(sig_info)
+
+        # adding bip32_derivation to outputs
+        outputs_meta = []
+        for i, txo in enumerate(psbt.glob.unsigned_tx.outputs()):
+            info = self.get_output_info(txo.address)
+            outputs_meta.append(info)
+        psbt.update_outputs(outputs_meta)
+
+        # adding redeem_script to inputs
+        psbt.generate_redeem_scripts()
+
+    def add_input_info_to_all_inputs(self, tx):
+        if tx.is_complete():
+            return
+        for txin in tx.inputs():
+            self.add_input_info(txin)
+
+    def _can_sign(self, psbt: PSBT):
+        for i, inp in enumerate(psbt.input_sections):
+            txin = psbt.glob.unsigned_tx.inputs()[i]
+            address = txin['address']
+            if self.is_mine(address):
+                return True
+                # TODO:
+                txin['type'] = self.get_txin_type(address)
+                # segwit needs value to sign
+                if txin.get('value') is None and is_input_value_needed(txin):
+                    received, spent = self.get_addr_io(address)
+                    item = received.get(txin['prevout_hash'] + ':%d' % txin['prevout_n'])
+                    tx_height, value, is_cb = item
+                    txin['value'] = value
+                self.add_input_sig_info(txin, address)
+
+    def can_sign(self, tx: Union[PSBT, Transaction]):
         if tx.is_complete():
             return False
         # add info to inputs if we can; otherwise we might return a false negative:
-        tx.add_inputs_info(self)
+        if isinstance(tx, Transaction):
+            self.add_input_info_to_all_inputs(tx)  # though note that this is a side-effect
+        else:
+            self.add_info_to_psbt(tx)
+            if self._can_sign(tx):
+                return True
```

Comment:
watch-only wallets will get here as `_can_sign` seems not to be finished.
so e.g. xpub only wallet can click "sign" button in tx dialog

Reply:
it's ok now?